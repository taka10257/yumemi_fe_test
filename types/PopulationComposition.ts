export type PopulationComposition = {
  year: number
  value: number
}
