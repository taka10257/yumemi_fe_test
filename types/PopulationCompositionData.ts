export type PopulationCompositionData = {
  year: number
  value: number
}
