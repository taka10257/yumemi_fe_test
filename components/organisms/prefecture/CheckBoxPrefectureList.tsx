import React, { useEffect, useState } from 'react'
import { Prefecture } from '../../../types/Prefecture'
import CheckBoxPrefecture from '../../molecules/checkbox/prefecture/CheckBoxPrefecture'
import styles from './CheckBoxPrefectureList.module.sass'
import axios from 'axios'
import { ApiError } from '../../../types/pages/ApiError'

type CheckBoxPrefectureListProps = {
  onCheck(prefCode: number, prefName: string): void
  onUnCheck(prefCode: number): void
}

export default function CheckBoxPrefectureList({
  onCheck,
  onUnCheck,
}: CheckBoxPrefectureListProps) {
  const [prefectures, setPrefectures] = useState<Prefecture[]>([])
  const [message, setMessage] = useState<string>('都道府県データの取得中...')

  useEffect(() => {
    if (prefectures.length === 0) {
      getPrefectures()
        .then((apiResult) => {
          setPrefectures(apiResult)
        })
        .catch((err: ApiError) => {
          setMessage(err.errorMessage)
        })
    }
  })

  return (
    <div>
      {prefectures.length === 0 ? (
        <>{message}</>
      ) : (
        <div>
          <div className={styles.title_block}>
            <p>都道府県を選択してください</p>
          </div>
          <div className={styles.checkbox_list}>
            {prefectures.map((prefecture) => (
              <div key={prefecture.prefCode} className={styles.checkbox}>
                <CheckBoxPrefecture
                  onCheck={onCheck}
                  onUnCheck={onUnCheck}
                  {...prefecture}
                />
              </div>
            ))}
          </div>
        </div>
      )}
    </div>
  )
}

function getPrefectures(): Promise<Prefecture[]> {
  return new Promise<Prefecture[]>((resolve, reject) => {
    axios
      .get<Prefecture[] | ApiError>('/api/prefectures')
      .then((result) => {
        const dataList = result.data as Prefecture[]
        if (!dataList || dataList.length === 0) {
          reject({ errorMessage: 'エラーが発生しました' })
          return
        }
        resolve(dataList)
      })
      .catch((err: ApiError) => {
        console.log(err)
        reject(err)
      })
  })
}
