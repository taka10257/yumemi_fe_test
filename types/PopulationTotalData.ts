import { PopulationCompositionData } from './PopulationCompositionData'

export type PopulationTotalData = {
  prefCode: number
  prefName: string
  data: PopulationCompositionData[]
}
