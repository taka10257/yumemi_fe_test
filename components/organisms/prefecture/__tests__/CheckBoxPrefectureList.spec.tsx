import React from 'react'
import { shallow } from 'enzyme'
import { expect } from '@jest/globals'
import CheckBoxPrefectureList from '../CheckBoxPrefectureList'
import axios from 'axios'
import { Prefecture } from '../../../../types/Prefecture'

jest.mock('axios')

describe('components/organisms/prefecture/CheckBoxPrefectureList', () => {
  describe('正常系', () => {
    it('初期表示', () => {
      ;(axios.get as jest.Mock).mockResolvedValue({
        data: testData(),
      })
      const checkboxList = shallow(
        <CheckBoxPrefectureList onCheck={jest.fn} onUnCheck={jest.fn} />
      )
      expect(checkboxList).toMatchSnapshot()
    })

    // TODO: useEffect利用時のテストを動かす
    it('都道府県データ更新した状態', async () => {
      // const checkboxList = shallow(
      //   <CheckBoxPrefectureList onCheck={jest.fn} onUnCheck={jest.fn} />
      // )
      //
      // // MEMO: このコードを入れることでstateが更新される(なぜ動くのかあまりわかってない)
      // //   参考: https://www.fixes.pub/program/325238.html
      // await act(async () => {
      //   await new Promise((resolve) => setTimeout(resolve, 0))
      // })
      //
      // expect(checkboxList).toMatchSnapshot()
    })
  })
})

function testData(): Prefecture[] {
  return [
    { prefCode: 1, prefName: '北海道' },
    { prefCode: 2, prefName: '青森県' },
  ]
}
