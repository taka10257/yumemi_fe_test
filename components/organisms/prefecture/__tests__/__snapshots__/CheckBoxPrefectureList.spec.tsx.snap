// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/organisms/prefecture/CheckBoxPrefectureList 正常系 初期表示 1`] = `
<div>
  都道府県データの取得中...
</div>
`;
